clang/lib/CodeGen/CGDebugInfo.cpp

```diff
@@ -5401,6 +5409,8 @@ std::string CGDebugInfo::GetName(const Decl *D, bool Qualified) const {
             // feasible some day.
             return TA.getAsIntegral().getBitWidth() <= 64 &&
                    IsReconstitutableType(TA.getIntegralType());
+          case TemplateArgument::StructuralValue:
+            return false;
```

Comment:
@dwblaikie, could you verify please? Maybe, non-type template arguments of floating point types are reconstitutable though?

Reply:
Ping @dwblaikie

Reply:
I don't think this is blocking either way, at worse we have suboptimal debugging experience for a new feature

Reply:
Sorry for the delays in the debug info review for this - it is on my list :/

Reply:
I did finally take a look at this - I think the Clang side of things is fine - if anything improvements could be made to LLVM to be able to lower constants to DWARF for a wider variety of values.
eg: For the `float` example, the IR is:
```
!8 = !DITemplateValueParameter(name: "F", type: !9, value: float 1.000000e+00)
```
But the DWARF backend can't handle the float constant - the handling is in `DwarfUnit::constructTemplateValueParameterDIE` (in llvm/lib/CodeGen/AsmPrinter/DwarfUnit.cpp) and currently only handles ConstantInt (for basic stuff - bools, ints, etc) and GlobalValue (for pointer non-type template parameters).

So for these new floats and such, the DWARF doesn't include the value, only the type.

GCC uses this encoding, for instance:
```
0x0000002b:     DW_TAG_template_value_parameter [3]   (0x0000001e)
                  DW_AT_name [DW_FORM_string]   ("F")
                  DW_AT_type [DW_FORM_ref4]     (cu + 0x004d => {0x0000004d} "float")
                  DW_AT_const_value [DW_FORM_block1]    (<0x04> 00 00 80 3f )
```

Reply:
Thanks for clarifying! Probably, I'll come back to this some day to extend the backend, but I'm not sure.